Overhaul the hooks chapter (#2899)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/Events/Hooks/Index.rst

Hooks
=====

> Hooks are basically places in the source code where a user function will be
> called for processing, if such has been configured. While there are conventions

Reviewer: ```suggestion
Hooks are basically places in the source code where a user function will be
called for processing, if such has been configured. The user function is executed
as part of the method where the hook resides, nevertheless the foreign function
has usually a very limited context and can only add or change distinct variables.
While there are conventions
```
The last sentence is not clear for me.

Author: How would you phrase this?

Hooks
=====

Hooks are basically places in the source code where a user function will be
called for processing, if such has been configured. While there are conventions
and best practises of how hooks should be implemented the hook concept itself
does not prevent it from being used in any way.

> Hooks are being phased-out in the TYPO3 Core and extension authors are also

Reviewer 2: Hmm, that does not really help as most hooks don't have a corresponding event (apart from the deprecated ones from v12.1-v12.4).

Author: As far as I understand almost all Hooks outside of the datahandler have become deprecated or removed in the Core. And what is wrong about advising extension authors to introduce events and not news Hooks?

Author: I tried to explain it more clear

..  literalinclude:: _ext_localconf_addhook.php
    :language: php
    :caption: EXT:site_package/ext_localconf.php

This hook registers the class/method name to a hook inside of
:php:`\TYPO3\CMS\Core\DataHandling\DataHandler`. The hook calls the user
function after the cache has been cleared. The user function
will receive parameters which allows it to see what clear-cache action was
performed and typically also an object reference to the parent object. Then the
user function can take additional actions as needed.

The class has to follow the PSR-4 class name scheme to be available in
:ref:`autoloading <autoloading_classes>`.

If we take a look inside of :code:`\TYPO3\CMS\Core\DataHandling\DataHandler` we
find the hook to be activated like this:

..  literalinclude:: _DataHandler.php
    :language: php
    :caption: :code:`\TYPO3\CMS\Core\DataHandling\DataHandler` (excerpt)

This is how hooks are typically constructed. The main action happens in line 5
where the function :code:`\TYPO3\CMS\Core\Utility\GeneralUtility::callUserFunction()`
is called. The user function is called with two arguments, an array with
variable parameters and the parent object.

> In line 24 the contents of the parameter array is prepared. This is of

Reviewer: ```suggestion
In line 24 the content of the parameter array is prepared. This is of
```
else it had to be "are prepared"

Using hooks
===========

The two lines of code below are an example of how a hook can be used for
clear-cache post-processing. The objective of this could be to perform
additional actions whenever the cache is cleared for a specific page:

..  literalinclude:: _ext_localconf_addhook.php

>     :language: php

Reviewer 2: language is not really necessary as it is the default. But is does not hurt.

Author: Just like with the code-blocks we should always use a language. This way it is more transparent how to make snippets for other languages. Also we decided not to rely on defaults.

    :caption: EXT:site_package/ext_localconf.php

This hook registers the class/method name to a hook inside of
:php:`\TYPO3\CMS\Core\DataHandling\DataHandler`. The hook calls the user

> function after the `clear-cache` command has been executed. The user function

Reviewer 2: Hmm, can we write "after the cache has been cleared" to avoid confusion with commands?

naming scheme should be used:

.. index::
   Hooks;  $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']
   pair: Hooks; Extensions
.. _hooks-extensions:

$GLOBALS['TYPO3\_CONF\_VARS']['EXTCONF']
----------------------------------------

**Configuration space for third-party extensions.**

This will contain all kinds of configuration options for specific
extensions including possible hooks in them! What options are
available to you will depend on a search in the documentation for that
particular extension.

..  literalinclude:: _ext_localconf_schema.php
    :language: php
    :caption: EXT:some_extension/ext_localconf.php

`<extension_key>`
    The unique extension key

`<sub_key>`
    Whatever the script defines. Typically it identifies
    the context of the hook

`<value>`
    It is up to the extension what the values mean, if they
    are mere configuration options or hooks or whatever and how deep the
    arrays go. Read the source code where the options are implemented to
    see. Or the documentation of the extension, if available.

..  note::

    :php:`$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']` was the recommended place where to

>     put hook configurations inside third party extensions. It is not recommended anymore

Reviewer 2: ```suggestion
    put hook configurations inside third-party extensions. It is not recommended anymore
```

Creating hooks
==============

..  note::

>     It is highly recommended to dispatch `Psr-14 events

Reviewer 2: ```suggestion
    It is highly recommended to dispatch `PSR-14 events
```

    <EventDispatcherQuickStartDispatching>` instead of introducing new hooks.
    Existing hooks should be migrated to events.

There are two main methods of calling a user defined function in
TYPO3.

:php:`\TYPO3\CMS\Core\Utility\GeneralUtility::callUserFunction()`
    The classic way. Takes a
    file/class/method reference as value and calls that function. The
    argument list is fixed to a parameter array and a parent object. So
    this is the limitation. The freedom is that the reference defines the
    function name to call. This method is mostly useful for small-scale
    hooks in the sources.

:php:`\TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance()`

>     Create an object from a user defined

Reviewer 2: ```suggestion
    Create an object from a user-defined
```

    file/class. The method called in the object is fixed by the hook, so

>     this is the non-flexible part. But it is cleaner in other ways, in

Reviewer 2: Two times "cleaner" in the sentence. Can we avoid it?

"But it is cleaner ... makes the API cleaner".

    particular that you can even call many methods in the object and you
    can pass an arbitrary argument list which makes the API cleaner.
    You can also define the objects to be singletons,
    instantiated only once in the global scope.

Here are some examples:

.. _hooks-creation-object:

Using `\TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance()`
--------------------------------------------------------------

Data submission to extensions:

> ..  code-block:: php

Reviewer 2: Why not moved into a separate file?

    :caption: EXT:some_extension/Classes/SomeClass.php

    use TYPO3\CMS\Core\Utility\GeneralUtility;

    // Hook for processing data submission to extensions
    foreach ($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['my_custom_hook']
         ['checkDataSubmission'] ?? [] as $className) {
      $_procObj = GeneralUtility::makeInstance($className);
      $_procObj->checkDataSubmission($this);
    }

.. _hooks-creation-function:

Using with `\TYPO3\CMS\Core\Utility\GeneralUtility::callUserFunction()`
-----------------------------------------------------------------------

Constructor post-processing:

.. code-block:: php
   :caption: EXT:some_extension/Classes/SomeClass.php

>    use TYPO3\CMS\Core\Utility\GeneralUtility;

Reviewer 2: Why not moved into a separate file?

Hook configuration
==================

> Most hooks in the TYPO3 Core have been converted into events which are

Reviewer 2: ```suggestion
Most hooks in the TYPO3 Core have been converted into PSR-14 events which are
```

completely listed in the :ref:`event list <eventlist>`.

There is no complete index of the remaining hooks in the Core. The following
naming scheme is usually used by third-party extensions and the remaining

> Core hooks:

Reviewer 2: Why a colon? Hooks from core are mostly available under SC_OPTIONS not EXTCONF?

Author: "third-party extensions and the remaining Core hooks:"

$GLOBALS['TYPO3\_CONF\_VARS']['SC\_OPTIONS']
============================================

**Configuration space for Core extensions.**

This array is created as an ad hoc space for creating hooks from any
script. This will typically be used from the Core scripts of TYPO3
which do not have a natural identifier like extensions have their
extension keys.

..  literalinclude:: _ext_localconf_schema.php
    :language: php
    :caption: EXT:some_extension/ext_localconf.php

`<main_key> `
    The relative path of a script (for output scripts it
    should be the "script ID" as found in a comment in the HTML header )

`<sub_key>`

>     Whatever the script defines. Typically it identifies
>     the context of the hook.

Reviewer: ```suggestion
    This is defined by the script. Typically it identifies
    the context of the hook.
```

$GLOBALS['TYPO3\_CONF\_VARS']['SC\_OPTIONS']
============================================

**Configuration space for Core extensions.**

This array is created as an ad hoc space for creating hooks from any
script. This will typically be used from the Core scripts of TYPO3
which do not have a natural identifier like extensions have their
extension keys.

..  literalinclude:: _ext_localconf_schema.php
    :language: php
    :caption: EXT:some_extension/ext_localconf.php

`<main_key> `
    The relative path of a script (for output scripts it
    should be the "script ID" as found in a comment in the HTML header )

`<sub_key>`
    Whatever the script defines. Typically it identifies
    the context of the hook.

`<index>`

>     Integer index typically. Can be unique string if you have

Reviewer 2: ```suggestion
    Integer index typically. Can be a unique string, if you have
```

    a reason to use that. Normally it has no greater significance since
    the value of the key is not used. The hooks normally traverse over the
    array and uses only the value (function reference)

`<function_reference>`
    A function reference using the syntax of
    :php:`\TYPO3\CMS\Core\Utility\GeneralUtility::callUserFunction()` as a function
    or :php:`\TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance()` as a class name
    depending on implementation of the hook.

>     A namespace function has the quoted string format :php:`'Foo\\Bar\\MyClassName->myUserFunction'`

Reviewer 2: I think we should only mention the ::class thingy as best practise (can be resolved by IDEs),

    or a format using an unquoted class name :php:`\Foo\Bar\MyClassName::class . '->myUserFunction'`.

>     The latter is available since PHP 5.5.

Reviewer 2: PHP 5.5, really? I suggest to throw that away,

>     A namespace class name can be in the FQCN quoted string format :php:`'Foo\\Bar\\MyClassName'`,

Reviewer 2: Here too.
